text/0000-promote-aarch64-unknown-linux-gnu-to-tier1.md

```diff
@@ -0,0 +1,188 @@
+
+ - Software stacks built for those domains predominantly use an AArch64 Linux kernel build.
+
+ - Rust presents an attractive value proposition across all such domains, irrespective of the underlying processor architecture.
+
+ - **As such, the Rust aarch64-unknown-linux-gnu target's ecosystem presents very strong viability for the long term.**
+
+**2.b. The long term viability of supporting the target should be clear.**
+
+ - It is hard to concretely quantify this aspect.
+
+ - It is worth calling out, in the same vein as the previous point, that given the increasing prevalance of AArch64 silicon deployments and given Rust's general value proposition, **supporting the Rust aarch64-unknown-linux-gnu target presents very strong viability for the long term.**
+
+**2.c. The target must have substantial and widespread interest within the Rust developer community.**
+
+ - It is hard to concretely quantify this aspect.
+
+ - It is generally fair to state that **there is already substantial and widespread interest for the aarch64-unknown-linux-gnu target in the Rust developer commmunity**.
+
+ - It is also generally fair to state that there is a clear upward trend in the use of AArch64 systems as self hosted development environments. 
+
+ - Most major operating system environments support hosted development on AArch64 based systems and this trend is increasing.
+
+ - As a somewhat related note: Slow but steady progress is being made to support Windows AArch64 targets, initially for cross-platform development. This shall inevitably trend towards hosted development.
+
+ - As such, **it is very likely that developer interest in Rust on aarch64-unknown-linux-gnu will continue to increase in the medium to long term.**
+
+**2.d. The target must serve the interests of multiple production users of Rust across multiple organizations or projects.**
+
+ - It is hard to concretely quantify this aspect.
+
+ - Most major Arm software ecosystem partners are either already using Rust extensively, or are building up to extensive use. A few publically known examples are Microsoft and Google. There are many more.
+
+ - Arm itself recognises Rust as an important component to consider in a broader horizontal safety and security foundation across multiple processor portfolios. 
+
+ - Arm has dedicated a small team to help improve Rust for the aarch64-unknown-linux-gnu target.
+
+ - **It is very likely that support for aarch64-unknown-linux-gnu in these organisations will trend upwards commensurate with the increasing prevalence of AArch64 silicon based systems.**
+
+Points 3.a through 3.c from the [Guide-level explanation](#Guide-level-explanation) section above are addressed in the [Unresolved questions](#unresolved-questions) section below.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+**There is no drawback envisioned in promoting the Rust aarch64-unknown-linux-gnu to Tier-1.**
```

Comment:
There is one important caveat/drawback/? compared to other current T1 targets – most developers won’t yet have access to aarch64 hardware that's capable enough to support compiler development. We do not currently have any infrastructure to provide such access either. As such there's a significant limitation in terms of developer-power that we would be able to mobilize in case something broke on aarch64.

That’s not an explicit requirement for a target to become T1 though (should it be?), so not saying this is a blocker or anything of the sort.

Reply:
Within a year or two it will likely become more reasonable to use the latest raspberry pi 4 (the 8gb model) as a compiler development system. Currently the 64-bit support of rpi4 is a little on the "beta" side, but folks are working on it. Both with the official raspberry pi OS and also using off the shelf linux distros like ubuntu and arch.

Maybe instead of "reasonable" I should say "possible". 1.5Mhz and 8GB of RAM means you can do `cargo check` and see results in "minutes", but a full build/test of the compiler would very likely take "hours". Might have to let CI occasionally do some of the heavy work.

source: I use a 4gb rpi4 for development of smaller projects and it's "slower than a proper desktop" speed, but you can still get work done.

Reply:
Another possibility would be to use a cloud based aarch64 system like the one linked to in the RFC ([Packet c2-large-arm](https://www.packet.com/cloud/servers/c2-large-arm/)) which seems reasonably priced in comparison to x86_64 options.

Reply:
Or AWS's Graviton CPUs, which are heavily subsidized and are actually priced below the x86_64 CPUs with equivalent performance

Reply:
I don’t think your average contributor will be willing to spend their own dollars just to hack on an architecture that they don’t have much stake in.

That said, this doesn't really affect me much, I've in my possession of a fairly competent aarch64 machine already. If others don't consider it a problem, that’s fine by me.

Reply:
> There is one important caveat/drawback/? compared to other current T1 targets – most developers won’t yet have access to aarch64 hardware that's capable enough to support compiler development. We do not currently have any infrastructure to provide such access either. As such there's a significant limitation in terms of developer-power that we would be able to mobilize in case something broke on aarch64.
> 
> That’s not an explicit requirement for a target to become T1 though (should it be?), so not saying this is a blocker or anything of the sort.

Hi @nagisa.

Thanks for your comment. You raise a good point. 

I think that it is reasonable to state that at present access to AArch64 systems capable of typical self hosted development is not as widespread as it perhaps needs to be. 

That said, the key points to consider are:

1. That _adequate_ access has been enabled such that the essential Rust lang compiler tests can be reliably run. Something that is key to track the evolution of the Rust lang master repository on AArch64 systems.

2. That there is strong evidence of rapid commoditisation of AArch64 silicon based systems that are convenient for self hosted development.

3. That until that commoditisation is truly widespread, the organisations that are putting skin into the game will likely provide suitable stop gaps.

On the basis of the above points, I am sincerely of the opinion that this problem should not impede the intent here. In fact, moving to Tier-1 is likely to accelerate the commoditisation of AArch64 systems to an extent that suitable systems are genuinely accessible to the Rust developer community.

I hope that the above assuages your concern. Happy to discuss further through to a mutually agreeable conclusion.

Thanks

Reply:
> Within a year or two it will likely become more reasonable to use the latest raspberry pi 4 (the 8gb model) as a compiler development system. Currently the 64-bit support of rpi4 is a little on the "beta" side, but folks are working on it. Both with the official raspberry pi OS and also using off the shelf linux distros like ubuntu and arch.
> 
> Maybe instead of "reasonable" I should say "possible". 1.5Mhz and 8GB of RAM means you can do `cargo check` and see results in "minutes", but a full build/test of the compiler would very likely take "hours". Might have to let CI occasionally do some of the heavy work.
> 
> source: I use a 4gb rpi4 for development of smaller projects and it's "slower than a proper desktop" speed, but you can still get work done.

Hi and thank you for your comments!

You make good points about how currently available off the shelf AArch64 silicon provides some measure of respite for self hosted aarch64-unknown-linux-gnu development. While this is not 'perfect', rest assured that more capable AArch64 silicon targeting Linux is certain and certain in a reasonable time frame.

Thanks.

Reply:
> Or AWS's Graviton CPUs, which are heavily subsidized and are actually priced below the x86_64 CPUs with equivalent performance

Hi and thanks for your comments.

Agree that AWS Graviton2 systems are good options. FYI: We mean to enable access to AWS Graviton2 instances for the Rust lang core team in the medium term.

Thanks!

Reply:
> I don’t think your average contributor will be willing to spend their own dollars just to hack on an architecture that they don’t have much stake in.
> 
> That said, this doesn't really affect me much, I've in my possession of a fairly competent aarch64 machine already. If others don't consider it a problem, that’s fine by me.

Hi again @nagisa.

I think the reality shall be that the non-average contributor won't have to spend their own dollars and that should be enough to gain the necessary critical mass in order to get a bulk of the essential support in motion.

That will then end up facilitating the average contributor as a side effect. Rest assured that commoditised access to suitable AArch64 silicon (for the aarch64-unknown-linux-gnu target) is inevitable in the short term.

The short term isn't now, I agree. However, that doesn't change the fact that the essential asks for a Tier-1 target aren't resolved today.

I hope that helps. Happy to discuss further.

Thanks.

Reply:
@raw-bin just to chime in here, I am *currenty* building a pi cluster for compilation as a service to offload some of the work on my PC, and I know at least two other developers doing the same thing

Reply:
Using https://github.com/mozilla/sccache if anyone is interested

Reply:
Don't forget about Apple's forthcoming ARM Macs, which were shown running AArch64 Linux under hardware virtualization.

Reply:
> I don’t think your average contributor will be willing to spend their own dollars just to hack on an architecture that they don’t have much stake in.

I don't know about your average contributor, but I've spun up EC2 instances before to answer some question on StackOverflow about an OS or ditro I didn't have on hand. Spinning up an instance for 15 minutes, or even an hour or two at a time, doesn't really cost enough to be worth worrying about, unless you have to do it regularly. How often does the average Rust contributor have to spend tracking down an architecture-specific bug when test fails?

That said, it would be nice if there were easier access to appropriate hardware or VMs for more contributors for this platform so more contributors could easily help out with regressions, or figure out why their PR caused an architecture-specific regression.

> FYI: We mean to enable access to AWS Graviton2 instances for the Rust lang core team in the medium term.

What team specifically? The core team is probably not the appropriate one; but if access were given to several of the relevant [teams](https://github.com/rust-lang/team/tree/master/teams), such as the compiler team, compiler contributors, libs team, and probably a few other teams that I haven't listed, then that would probably give access to a large enough pool of regular contributors who are likely to have to deal with arch specific issues, and address the concerns about contributors not having access.

Reply:
Hi @lambda. 

Good points, thanks. 

Agree that facilitating access to relevant teams will help.

The way I anticipate that this will work out (on the basis of the discussions I’ve had thus far and the outcomes from them) is that as and when systems are needed the core team will reach out and I’ll help with provisioning and access etc. 

When we started out the immediate need was getting CI sorted so that was the primary focus. The discussions for understanding the ask and working out specifics came out of the core team. I expect that this will then get handed over to the infrastructure team perhaps. So a kind of triaging protocol. 

Incidentally we ended up enabling WASM work similarly. 

I think this can and will be done similarly in future.

Reply:
I've added a commit to the PR that addresses comments made by @lambda.

I think I've responded suitably to the other good comments.

I'll now resolve this conversation. Happy to engage further if needed.

Reply:
Yep, CI is absolutely the top priority, and probably the only one that would need to block promotion; if all else fails, pushing branches to be tested by CI can work in a pinch for those who don't have access to hardware for debugging. I'd put expanded access for more contributors as a high-priority request, not a blocker.

I think that between cloud systems being readily available, Raspberry Pi 4 and other single-board computers being easily available and now powerful enough to build Rust, Chromebooks, and the upcoming ARM Macs, there should be sufficient access to this architecture for a wide enough number of contributors for this not to be a blocking concern.